### PR TITLE
ArC - support @TransientReference

### LINKED ISSUE #6858: ArC - support @TransientReference
See https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#dependent_destruction for more information.

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #6858

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java [hunk 1]
```diff
@@ -1414,6 +1462,12 @@ void implementCreateForClassBean(ClassOutput classOutput, ClassCreator beanCreat
                 create.invokeVirtualMethod(MethodDescriptor.of(methodInjection.target.asMethod()), instanceHandle,
                         referenceHandles);
             }
+
+            // Destroy injected transient references
+            for (TransientReference transientReference : transientReferences) {
```
Comment (manovotn) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:1467]: Maybe extract to a util method? The same destroy logic appears four times in this class.

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    static class BeerProducer {
+
+        static final AtomicInteger COUNTER = new AtomicInteger();
+        static final List<Integer> DESTROYED = new CopyOnWriteArrayList<>();
+
+        @Dependent
+        @Produces
+        Beer newBeer(InjectionPoint injectionPoint) {
+            int id;
+            if (injectionPoint.getAnnotated() instanceof AnnotatedField) {
```
Comment (manovotn) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java:112]: Nice, I didn't even know we actually have `Annotated` impls just for `InjectionPoint` :)

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.TransientReference;
+import javax.enterprise.inject.spi.AnnotatedField;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class TransientReferenceDestroyedTest {
```
Comment (manovotn) [inline independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java:25]: Should we also test injection of transient reference bean into observer method?

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java (hunk 2 above)
Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java:112]: Yeah, it's ugly I know but we need to support some old stuff ;-)

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java (hunk 3 above)
Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java:25]: That's a good point. In fact, we don't support them ATM ;-). I'll do that.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java (hunk 1 above)
Comment (mkouba) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java:1467]: Well, it's just a very simple iteration and  the util method would have to accept three params. But you're probably right ;-)

File: independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java (hunk 3 above)
Comment (mkouba) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java:25]: Wait a minute, for observers the injected dependent beans are destroyed even without `@TransientReference`...

Comment (manovotn) [reply independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/transientreference/TransientReferenceDestroyedTest.java:25]: Hah, a morning intelligence test (in which I apparently failed) :-D

**Review by manovotn — APPROVED** [review]

**Pull request merged**